osdconfig/api.go

```diff
@@ -0,0 +1,337 @@
+func (manager *configManager) GetContext() context.Context {
+	return manager.ctx
+}
+
+// GetStatus returns execution status
+func (manager *configManager) GetStatus() map[string]*Status {
+	manager.Lock()
+	M := make(map[string]*Status)
+	for key, val := range manager.status {
+		key := key
+		M[key] = new(Status)
+		M[key].Err = val.Err
+		M[key].Duration = val.Duration
+	}
+	manager.Unlock()
+	return M
+}
+
+// Error returns error status of callback execution
+func (manager *configManager) Error() error {
+	manager.Lock()
+	defer manager.Unlock()
+	if manager.status == nil {
+		return nil
+	}
+	for _, val := range manager.status {
+		if val != nil {
+			if val.Err != nil {
+				return ExecErr
+			}
+		}
+	}
+	return nil
+}
+
+// Close closes the contexts and releases bookkeeping memory
+func (manager *configManager) Close() {
+	// cancel contexts and wait for any pending tasks to complete
+	manager.runCancel()
+	manager.cancel()
+	manager.Wait()
+
+	// clean up resources
+	manager.cc = nil
+	manager.cb = nil
+	manager.opt = nil
+}
+
+// Abort sends context cancellation to running callbacks if any
+func (manager *configManager) Abort() {
+	logrus.Info("aborting via context cancellation")
+	manager.runCancel()
+	manager.Lock()
+	logrus.Info("abort successful")
+	manager.Unlock()
+}
+
+// Wait waits for running callbacks to finish
+func (manager *configManager) Wait() {
+	manager.Lock()
+	manager.Unlock()
+}
+
+// GetClusterConf retrieves cluster level data from kvdb
+func (manager *configManager) GetClusterConf() (*ClusterConfig, error) {
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(filepath.Join(baseKey, clusterKey))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, DataErr
+	}
+
+	config := new(ClusterConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetClusterConf sets cluster config in kvdb
+func (manager *configManager) SetClusterConf(config *ClusterConfig) error {
+	manager.Lock()
+	defer manager.Unlock()
+
+	if config == nil {
+		return InputErr
+	}
+
+	// push into kvdb
+	_, err := manager.cc.Put(filepath.Join(baseKey, clusterKey), config, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetNodeConf retrieves node config data using nodeID
+func (manager *configManager) GetNodeConf(nodeID string) (*NodeConfig, error) {
+	if len(nodeID) == 0 {
+		return nil, InputErr
+	}
+
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(getNodeKeyFromNodeID(nodeID))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, DataErr
+	}
+
+	config := new(NodeConfig)
+	if err = json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetNodeConf sets node config data in kvdb
+func (manager *configManager) SetNodeConf(config *NodeConfig) error {
+	manager.Lock()
```

Comment:
yes these can potentially go, but need more experiments. i actually put those in yesterday because i did not want people to fire multiple kvdb puts without regard for completion of callbacks. lock here only guarantees that either kvdb is being set or callbacks are running, not both at the same time.